### PR TITLE
Feature shylubasker interface

### PR DESCRIPTION
**Pull request by ndellingwood**
Work toward adding ShyLUBasker interface to Amesos2; both versions
of Basker can now be enabled simultaneously

Add explicit Amesos2::ShyLUBasker interface and remove experimental workaround
requiring use of flag SHYLU_NODEBASKER to CMake flags

Use of ShyLUBasker in Amesos2 will require the following:
1. Trilinos_ENABLE_ShyLU_NodeBasker:BOOL=ON
2. Amesos2_ENABLE_ShyLU_NodeBasker:BOOL=ON
3. CMAKE_CXX_FLAGS:STRING="-DSHYLU_NODEBASKER" is no longer needed

Use of Basker in Amesos2 requires no changes.

@trilinos/amesos2 

## Description

Add explicit Amesos2::ShyLUBasker interface and remove experimental workaround
requiring use of flag SHYLU_NODEBASKER to CMake flags

Use of ShyLUBasker in Amesos2 will require the following:
1. Trilinos_ENABLE_ShyLU_NodeBasker:BOOL=ON
2. Amesos2_ENABLE_ShyLU_NodeBasker:BOOL=ON
3. CMAKE_CXX_FLAGS:STRING="-DSHYLU_NODEBASKER" is no longer needed

Use of Basker in Amesos2 requires no changes.

Amesos2::Basker: Renamed Basker class to BaskerClassic

Amesos2::Basker: Internal amesos2/src/basker code modified so that Basker class renamed to BaskerClassic, changes unseen by Amesos2 users. 

## Motivation and Context

CMake updates to ShyLU Basker address issue #2075 

## Related Issues

* Closes 
* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to #2075 
* Part of 
* Composed of 

## How Has This Been Tested?

Tested locally with gcc/5.4, scotch 6.0.3, OpenMP, with each combo of MPI ON/OFF and Complex ON/OFF. 
Auto-tester will add testing coverage. 

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [x] These changes break backwards compatibility.

## Additional Information

The ShyLU Basker changes break backward compatibility - please follow the configure steps mentioned in **Description** above.

### CODE REVIEW

**Comment by ndellingwood** [comment]
@hkthorn I think you'll be interested in this :)

**Comment by ndellingwood** [comment]
WIP added to title for now so autotester does not spam

**Comment by mhoemmen** [comment]
Feature request: Enable those CMake flags by default :-)

**Comment by srajama1** [comment]
@ndellingwood : I think we can make this feature request happen in two steps. It had lot of problems before. We have refactored Amesos, Amesos2 and Trilinos SS. Amesso2 depends on TrilinosSS now. We need to call AMD from TrilinosSS before calling ClassicBasker in the symbolic step of Amesos2 (or add a symbolic call in ClassicBasker and call AMD from there). ClassicBasker is robust enough with just this addition to be enable by default on host space.

The multithreaded basker is not robust enough yet to be enabled by default. We need to fix those nagging issues before enabling it.

**Comment by ndellingwood** [comment]
@srajama1 I updated this PR after the 2.7.00 Kokkos+KokkosKernels promotion. 
Regarding the BaskerClassic update request to add symbolic factorization (and then enable this solver by default), I like the second option you suggested best, "add a symbolic call in ClassicBasker and call AMD from there." From the AMD documentation, it returns a permutation vector but does not permute the matrix, so we will need to store each of these and I think the best way is for BaskerClassic to assume responsibility for the management of that data. This way, the Amesos2 impl stays consistent with the other solvers and does not need additional members for this data management. Also, BaskerClassic already has some routines for permuting vectors and matrices, while these would need to be added for Amesos2. 
If your okay with this route for the changes, I'd prefer to submit the BaskerClassic changes in a separate PR since they will touch the internals of the solver (rather than the Amesos2 BaskerClassic interface), and this will help me more easily triage potential issues that surface following either of these changes.

File: packages/shylu/shylu_node/basker/src/shylubasker_trilinos_decl.hpp [hunk 1]
```diff
@@ -1,22 +1,32 @@
-#ifndef BASKER_TRILINOS_DECL_HPP
-#define BASKER_TRILINOS_DECL_HPP
+#ifndef SHYLUBASKER_TRILINOS_DECL_HPP
+#define SHYLUBASKER_TRILINOS_DECL_HPP
+
+#if 0
```
Comment (srajama1) [inline packages/shylu/shylu_node/basker/src/shylubasker_trilinos_decl.hpp]: Can we remove this ?

File: packages/shylu/shylu_node/basker/src/shylubasker_order_btf.hpp [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
-#define BASKER_ORDER_BTF_HPP
+#ifndef SHYLUBASKER_ORDER_BTF_HPP
+#define SHYLUBASKER_ORDER_BTF_HPP
 
-#include "basker_types.hpp"
-#include "basker_sswrapper.hpp"
+#include "shylubasker_types.hpp"
+#include "shylubasker_sswrapper.hpp"
 
 #include "trilinos_btf_decl.h"
```
Comment (srajama1) [inline packages/shylu/shylu_node/basker/src/shylubasker_order_btf.hpp]: Check this is suitesparse

File: packages/shylu/shylu_node/basker/src/basker_order_bts.hpp [hunk 3]
```diff
@@ -1,11 +0,0 @@
-#ifndef BASKER_ORDER_BTS_HPP
```
Comment (srajama1) [inline packages/shylu/shylu_node/basker/src/basker_order_bts.hpp]: Check

File: packages/amesos2/example/SimpleSolveNonContigMap.cpp [hunk 4]
```diff
@@ -77,7 +77,11 @@
 int main(int argc, char *argv[]) {
   Teuchos::GlobalMPISession mpiSession(&argc,&argv);
 
+#ifdef HAVE_AMESOS2_SHYLUBASKER
```
Comment (srajama1) [inline packages/amesos2/example/SimpleSolveNonContigMap.cpp]: We can enable both of them, right ?

File: packages/amesos2/src/basker/basker.hpp [hunk 5]
```diff
@@ -42,7 +42,7 @@
 #define ASSERT(a) mxAssert(a,"")
 #define REALLOC(ptr,size) mxRealloc(ptr, size)
 #define CALLOC(num, size) mxCalloc(num, size)
```
Comment (srajama1) [inline packages/amesos2/src/basker/basker.hpp]: Can we rename all of these ?

**Comment by ndellingwood** [comment]
Thanks for review @srajama1 ! Recent commit addressed comments, will rebase on current develop branch and do some testing to make sure it is ready to remove [WIP]

File: packages/amesos2/src/basker/basker.hpp [hunk 6]
```diff
@@ -39,18 +39,18 @@
...
-#define REALLOC(ptr,size) mxRealloc(ptr, size)
-#define CALLOC(num, size) mxCalloc(num, size)
+#define BASKERASSERT(a) mxAssert(a,"")
+#define BASKERREALLOC(ptr,size) mxRealloc(ptr, size)
+#define BASKERCALLOC(num, size) mxCalloc(num, size)
 #define BASKERFREE(ptr)
 
 #else
 #include <assert.h>
 #include <stdlib.h>
```
Comment (mhoemmen) [inline packages/amesos2/src/basker/basker.hpp]: The `<xxx.h>` C standard library headers were deprecated in C++ as of C++98; please use the `<cxxx>` headers instead.

File: packages/amesos2/src/basker/basker_types.hpp [hunk 7]
```diff
@@ -34,18 +34,18 @@
...
-#define ASSERT(a) mxAssert(a,"")
-#define REALLOC(ptr,size) mxRealloc(ptr, size)
-#define CALLOC(num, size) mxCalloc(num, size)
+#define BASKERASSERT(a) mxAssert(a,"")
+#define BASKERREALLOC(ptr,size) mxRealloc(ptr, size)
+#define BASKERCALLOC(num, size) mxCalloc(num, size)
 #define BASKERFREE(ptr)
 
 #else
 #include <assert.h>
```
Comment (mhoemmen) [inline packages/amesos2/src/basker/basker_types.hpp]: See comment above.

**Comment by ndellingwood** [comment]
Removed WIP and letting autotester begin

**Comment by ndellingwood** [comment]
@srajama1 sorry to bother for another re-review, but I just pushed a final round of changes which should complete the PR. I mistakenly added an optimization to the serial Basker loadA_impl routine that isn't ready (I need to add a transpose solve to handle the CRS format, will pack that in with the symbolic stuff in a followup PR), replaced the deprecated C-standard headers, and fixed the unit tests in threaded ShyLUBasker to properly initialize Kokkos which fixed the increased runtime issue of the tests that we noticed.

**Comment by ndellingwood** [comment]
Pattern match PR #3094 for the ShyLUBasker tests regarding Tpetra changes.

**Comment by srajama1** [comment]
I don't know why this says Pre-merge inspection pending still.

**Comment by ndellingwood** [comment]
@srajama1 thanks! Maybe there was a bit of latency for the preinspection to catch up? I was able to merge though :)

**Comment by srajama1** [comment]
Ah, got it. Thanks !

**Pull request merged**